src/Misc/PerformanceViewer/performanceViewerCollector.ts

```diff
@@ -0,0 +1,225 @@
+const initialArraySize = 1800;
+
+// three octets in a hexcode. #[AA][BB][CC], i.e. 24 bits of data.
+const numberOfBitsInHexcode = 24;
+
+// Allows single numeral hex numbers to be appended by a 0.
+const hexPadding = "0";
+
+/**
+ * The collector class handles the collection and storage of data into the appropriate array.
+ * The collector also handles notifying any observers of any updates.
+ */
+export class PerformanceViewerCollector {
+    private _datasetMeta: Map<string, IPerfMetadata>;
+    private _strategies: Map<string, IPerfViewerCollectionStrategy>;
+    private _startingTimestamp: number;
+
+    /**
+     * Datastructure containing the collected datasets. Warning: you should not modify the values in here, data will be of the form [timestamp, numberOfPoints, value1, value2..., timestamp, etc...]
+     */
+    public readonly datasets: IPerfDatasets;
+    /**
+     * An observable you can attach to get deltas in the dataset. Subscribing to this will increase memory consumption slightly, and may hurt performance due to increased garbage collection needed.
+     */
+    public readonly datasetObservable: Observable<IPerfDataSliceSnapshot>;
+    /**
+     * An observable you can attach to get the most updated map of metadatas.
+     */
+    public readonly metadataObservable: Observable<Map<string, IPerfMetadata>>;
+
+    /**
+     * The offset for when actual data values start appearing inside a slice.
+     */
+    public static get SliceDataOffset() {
+        return 2;
+    }
+
+    /**
+     * The offset for the value of the number of points inside a slice.
+     */
+    public static get NumberOfPointsOffset() {
+        return 1;
+    }
+
+    /**
+     * Handles the creation of a performance viewer collector.
+     * @param _scene the scene to collect on.
+     * @param _enabledStrategyCallbacks the list of data to collect with callbacks for initialization purposes.
+     */
+    constructor(private _scene: Scene, _enabledStrategyCallbacks?: PerfStrategyInitialization[]) {
+        this.datasets = {
+            ids: [],
+            data: new DynamicFloat32Array(initialArraySize),
+            startingIndices: new DynamicFloat32Array(initialArraySize)
+        };
+        this._strategies = new Map<string, IPerfViewerCollectionStrategy>();
+        this._datasetMeta = new Map<string, IPerfMetadata>();
+        this.datasetObservable = new Observable();
+        this.metadataObservable = new Observable((observer) => observer.callback(this._datasetMeta, new EventState(0)));
+        if (_enabledStrategyCallbacks) {
+            this.addCollectionStrategies(_enabledStrategyCallbacks);
+        }
+    }
+
+    /**
+     * This method adds additional collection strategies for data collection purposes.
+     * @param strategyCallbacks the list of data to collect with callbacks.
+     */
+    public addCollectionStrategies(strategyCallbacks: PerfStrategyInitialization[]) {

```

Comment:
if I want to add a single one, do I need to send an array?
Maybe you could use spread operator (https://howtodoinjava.com/typescript/spread-operator/), or make this addCollectionStrategy while leaving this as a help function that iterates over an array and uses the single add method?

Reply:
Oh yeah thanks for the suggestion, I forgot I could simply just use the spread operator 😄

Reply:
Done!